### PR TITLE
hp::DoFHandler::execute_coarsening_and_refinement()

### PR DESCRIPTION
**Pull request by marcfehling**
This PR adds a new function to the `hp::DoFHandler` that sets new `active_fe_indices` according to refinement flags (#7777) and the registered hierarchy of finite elements (#7767). Tests verify implementations for both distributed and non-distributed triangulations.

Requested in #7515. Part of #3511.

### CODE REVIEW

File: source/hp/dof_handler.cc [hunk 1]
```diff
@@ -1506,6 +1506,30 @@ namespace hp
 
 
 
+  template <int dim, int spacedim>
+  void
+  DoFHandler<dim, spacedim>::execute_coarsening_and_refinement()
+  {
+    for (const auto &cell : active_cell_iterators())
+      if (cell->is_locally_owned())
```
Comment (marcfehling) [inline source/hp/dof_handler.cc]: Do we need to check for artificial cells? So far all tests pass without one...

File: include/deal.II/hp/dof_handler.h [hunk 2]
```diff
@@ -1111,9 +1124,6 @@ namespace hp
     /**
      * A function that will be triggered through a triangulation
      * signal just after the triangulation is modified.
-     *
-     * The function that restores the active_fe_flags of all cells that
-     * were refined.
```
Comment (marcfehling) [inline include/deal.II/hp/dof_handler.h]: The comment is no longer true: This function is no longer responsible for that.

File: source/hp/dof_handler.cc (hunk 1 above)
Comment (davydden) [reply source/hp/dof_handler.cc]: I don't think you need to set flags on `ghost` cells.

File: include/deal.II/hp/dof_handler.h [hunk 3]
```diff
@@ -479,6 +479,19 @@ namespace hp
...
+     * according to refinement and coarsening flags set and the hierarchy
+     * of the registered hp::FECollection.
+     *
+     * Since the current processor only has control over those cells it owns
+     * (i.e. the ones for which <code>cell-@>subdomain_id() ==
+     * this-@>locally_owned_subdomain()</code>), refinement and coarsening
+     * flags are only respected for those locally owned cells.
+     */
+    void
+    execute_coarsening_and_refinement();
```
Comment (davydden) [inline include/deal.II/hp/dof_handler.h]: maybe `execute_p_coarsening_and_refinement` ?

File: source/hp/dof_handler.cc (hunk 1 above)
Comment (marcfehling) [reply source/hp/dof_handler.cc]: Hmm. I guess I should extend the domain in those tests.

Right now there is always exactly one processor who owns the whole domain. The others just store one artificial cell on the top level.

Comment (marcfehling) [reply source/hp/dof_handler.cc]: Seems to work even with this extension. I guess we're good.

File: source/hp/dof_handler.cc [hunk 4]
```diff
@@ -1506,6 +1506,30 @@ namespace hp
 
 
 
+  template <int dim, int spacedim>
+  void
+  DoFHandler<dim, spacedim>::execute_coarsening_and_refinement()
```
Comment (davydden) [inline source/hp/dof_handler.cc]: Perhaps we should guide users a bit with this function.

what I did in `tests/hp/laplace.h` for https://github.com/dealii/dealii/pull/7513  was the following:

1. Mark cells for h-refinement using standard tools
2.selectively substituted some h-refinement flags for p-refinement
call triangulation.prepare_coarsening_and_refinement();
3. setup Solution Transfer and call
  soltrans.prepare_for_coarsening_and_refinement(solution_coarse);

4. **** increment fe_index *****
5. execute Refinement
  triangulation.execute_coarsening_and_refinement();
6. setup_system();
7. interpolate via solution transfer
  soltrans.interpolate(solution_coarse, solution);

Given that the sequence is not trivial in the presence of solution transfer, perhaps we can briefly document this in the function description. Maybe say something like

> If the solution transfer is to be done in the hp-FEM case, this function should be called after initialization of the solution transfer object and before calling `execute_coarsening_and_refinement()`.

Comment (marcfehling) [reply source/hp/dof_handler.cc]: Agreed. I'll extend the documentation for this function.

I guess a good place to give users a general overview over the functionality would be `doc/doxygen/headers/hp.h`. However, I would postpone that until the day we're happy with the interface (i.e. #7515 is done).

**Comment by marcfehling** [comment]
I forgot to add a changelog. I'll come up with that later.

File: include/deal.II/hp/dof_handler.h (hunk 3 above)
Comment (marcfehling) [reply include/deal.II/hp/dof_handler.h]: Good call. This would prevent confusion with the `Triangulation` equivalent.

**Comment by marcfehling** [comment]
Oh no! I thought about the `SolutionTransfer` workflow, which would fail in the current implementation for the `parallel::distributed` case: Right now, we update the `active_fe_indices` before we trigger the registered `pack_callback` functions. This should be the other way round.

The other cases ('sequential'/'parallel::shared') should be fine, since we only store the `dof_indices` on refined/coarsened cells to address their fraction of the old solution later (upon interpolation).

In the 'parallel::distributed' case, we prepare data to be transferred after we refine the `p4est` object, thus we can't just move the data transfer part somewhere else. My initial idea would be to introduce another signal to which all `hp::DoFHandlers` would connect their refinement execution. Thus, calling `tria.execute_coarsening_and_refinement()` with associated `hp::DoFHandlers` would always perform hp-refinement. I don't like this, but I don't see any alternative which would bloat up the whole code.

Pseudo-Code:
```
parallel::distributed::Triangulation::execute_coarsening_and_refinement():
  ...
  refine (parallel_forest);
  coarsen (parallel_forest);
  balance (parallel_forest);

  update_quadrant_cell_relations ();
  data_transfer.pack_data (pack_callbacks);
  repartition (parallel_forest);

  <<< PLACE NEW SIGNAL HERE >>>
  copy_local_forest_to_triangulation ();

  data_transfer.execute_transfer (parallel_forest);
  ...
```

To be consistent with the 'sequential'/'parallel::shared' cases, I would add the same signal to those as well. What do you think about this?

**Comment by marcfehling** [comment]
Oh no! This wouldn't work either, since we store all `active_fe_indices` with a `pack_callback` call.

I ran into a dead end and need to get back to the drawing board... :(

**Comment by marcfehling** [comment]
When we trigger the above `p_refinement` signal at the indicated position as pointed out above, we could use the still present `p_refine_flags` to already prepare the solution on the new mesh. This also has to happen for the transferred `active_fe_indices`.

I will make the `hp::DoFHandler::execute_p_coarsening_and_refinement()` function private, so that any kind of refinement has to be triggered via `Triangulation::execute_coarsening_and_refinement()`. Any other approach would be too tedious to handle and too confusing for the user.

Please flag this as WIP. I have the construction plans in mind, and will get on it tomorrow. Need to grab some sleep first :)

**Comment by marcfehling** [comment]
In fact, we neither need a new signal nor the `execute_p_refinement()` function. We could implement everything in the `pre_active_fe_index_transfer` and `post_active_fe_index_transfer` callback functions, in which we prepare the `active_fe_indices` according to the flags and set them afterwards.

**Pull request closed without merging**